table/hash_index.go

```diff
@@ -0,0 +1,94 @@
+package table
+
+import (
+	"encoding/binary"
+	"github.com/dgryski/go-farm"
+)
+
+const (
+	resultNoEntry  = 65535
+	resultFallback = 65534
+	maxRestart     = 65533
+)
+
+type hashIndexBuilder struct {
+	entries       []indexEntry
+	hashUtilRatio float32
+	invalid       bool
+}
+
+type indexEntry struct {
+	hash uint32
+
+	// blockIdx is the index of block which contains this key.
+	blockIdx uint16
+
+	// offset is the index of this key in block.
+	offset uint8
+}
+
+func newHashIndexBuilder(hashUtilRatio float32) hashIndexBuilder {
+	return hashIndexBuilder{
+		hashUtilRatio: hashUtilRatio,
+	}
+}
+
+func (b *hashIndexBuilder) addKey(key []byte, blkIdx uint32, offset uint8) {
+	if blkIdx > maxRestart {
+		b.invalid = true
+		b.entries = nil
+		return
+	}
+	h := farm.Fingerprint32(key)
+	b.entries = append(b.entries, indexEntry{h, uint16(blkIdx), offset})
+}
+
+func (b *hashIndexBuilder) finish(buf []byte) []byte {
+	if b.invalid || len(b.entries) == 0 {
+		return append(buf, u32ToBytes(0)...)
+	}
+
+	numBuckets := uint32(float32(len(b.entries)) / b.hashUtilRatio)
+	bufLen := len(buf)
+	buf = append(buf, make([]byte, numBuckets*3+4)...)
+	buckets := buf[bufLen:]
+	for i := 0; i < int(numBuckets); i++ {
+		binary.LittleEndian.PutUint16(buckets[i*3:], resultNoEntry)
+	}
+
+	for _, h := range b.entries {
+		idx := h.hash % numBuckets
+		bucket := buckets[idx*3 : (idx+1)*3]
+		blkIdx := binary.LittleEndian.Uint16(bucket[:2])
+		if blkIdx == resultNoEntry {
+			binary.LittleEndian.PutUint16(bucket[:2], h.blockIdx)
+			bucket[2] = h.offset
+		} else if blkIdx != h.blockIdx {
```

Comment:
Is it possible that the key is different, but the hash value and `blockIdx` is the same, only offset is different?

Reply:
It's possible, but I think it's ok to allow this. This type collision can be solved in `seekFromOffset`.

1. Navigate to this block and check key at `offset`
2. Key not the same
3. Seek in this block
4. Find the correct key

As you can see, hash index can still find correct result and eliminate a binary search in block index.